src/couch_replicator_rate_limiter.erl

```diff
@@ -0,0 +1,219 @@
+-module(couch_replicator_rate_limiter).
+-behaviour(gen_server).
+
+
+-include_lib("couch/include/couch_db.hrl").
+-include_lib("ibrowse/include/ibrowse.hrl").
+-include_lib("couch_replicator_api_wrap.hrl").
+
+
+-export([
+    start_link/0
+]).
+
+-export([
+    init/1,
+    terminate/2,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    code_change/3
+]).
+
+-export ([
+    maybe_start_rate_limiter/3,
+    maybe_decrement_rep_count/1,
+    maybe_delay_request/1
+    ]).
+
+-record(rep_limiter, {
+    requestCounter = 0,
+    lastUpdateTimestamp = 0,
+    pid,
+    replications = 0,
+    limit, % max requests
+    period % interval is in seconds
+}).
+
+
+%% For each unique host, an entry is a rep_limiter record
+-define(HOSTS, couch_replicator_limit_hosts).
+
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+
+update_host(Host, Limit, Period) ->
+    gen_server:call(?MODULE, {update_host, Host, Limit, Period}, infinity).
+
+
+decrement_replications(Host) ->
+    gen_server:call(?MODULE, {decrement_replications, Host}, infinity).
+
+
+% gen_server functions.
+init([]) ->
+    process_flag(trap_exit, true),
+    ets:new(?HOSTS, [set, public, named_table]),
+    {ok, nil}.
+
+
+handle_call({update_host, Host, Limit, Period}, _From, State) ->
+    Reps = case ets:insert_new(?HOSTS, {Host, #rep_limiter{}}) of
+        true ->
+            % Brand new host for replication, need to spawn
+            % a new resetter process and initialize
+            LoopPid = spawn_link(fun() ->
```

Comment:
Is it possible to not do this custom supervision? I'd like to err on the side of something generic, like an OFO or SOFO supervisor to handle a per-host rate-limiting proc. trap_exits are kinda code-smelly.